reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+- This proposal does not delegate ARIA attributes that aren't ID references, such as `aria-label`.
+
+### The `referenceDelegate` attribute
+
+A component can specify an element in its shadow tree to act as its "reference delegate". When the host component is the target of a IDREF like a label's `for` attribute, the delegate becomes the true target.
+
+The shadow root specifies the ID of the delegate element inside the shadow DOM. This is done either in JavaScript with the `referenceDelegate` attribute on the `ShadowRoot` object, or in HTML markup using the `shadowrootreferencedelegate` attribute on the `<template>` element.
+
+JavaScript example:
+
+```html
+<template id="t-fancy-input">
+  <input id="real-input" />
+</template>
+
+<script>
+customElements.define("fancy-input", 
+  class FancyInput extends HTMLElement {
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-fancy-input").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-input";
+    }
+  });
+</script>
+
+<label for="fancy-input">Fancy input</label>
+<fancy-input id="fancy-input"></fancy-input>
+```
+
+Equivalent with declarative shadow DOM:
+
+```html
+<label for="fancy-input">Fancy input</label>
+<fancy-input id="fancy-input">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegate="real-input">
+    <input id="real-input" />
+  </template>
+</fancy-input>
+```
+
+### The `referenceDelegateMap` attribute
+
+There are situations where it is necessary to delegate different reference types to different elements. For example, a listbox may want to delegate `aria-controls` to its root, and `aria-activedescendant` to one of the items inside the listbox.
+
+The `ShadowRoot.referenceDelegateMap` attribute allows for specifying elements based on the attribute that is being used to reference the host.
+
+The equivalent declarative attribute is `shadowrootreferencedelegatemap`, which is a comma-separated list of attribute to ID mappings.
+
+```html
+<input role="combobox"
+       aria-controls="fancy-listbox"
+       aria-activedescendant="fancy-listbox" />
+<fancy-listbox id="fancy-listbox">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegatemap="aria-controls: real-listbox,
+                                            aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
+
+Declaring the mappings using the JavaScript API would look like:
+```js
+this.shadowRoot_.referenceDelegateMap['aria-controls'] = 'real-listbox';
+this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-2';
```

Comment:
As `aria-activedescendant` would need to point to different IDs overtime, should we be explicit here, in preparation for spec text about this being a live reference in some way?

```js
this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-2';
// ...then later...
this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-1';
// ...etc.
```

Reply:
Yes, the intent is for these to be live references. I'll write that in explicitly.

Reply:
Bikeshedding a bit, but it feels more natural to me in JS to use the camelCase format (ala the `ARIMixin`):

```js
this.shadowRoot_.referenceDelegateMap.ariaActiveDescendant = 'option-2';
```

(You might think `ReferenceDelegateMap` could just extend `ARIAMixin`, but that would require first solving https://github.com/w3c/aria/issues/1732, since `ariaActiveDescendant` and `ariaLabelledBy` and friends don't currently exist. 🙂 So maybe it's best to just define the 8 IDref camelCase formats just for `ReferenceDelegateMap`.)

Reply:
`for` is also a funky case when it comes to the casing... Presumably this should be [`htmlFor`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLLabelElement/htmlFor), i.e.  `referenceDelegateMap.htmlFor`?